src/main/kotlin/org/jetbrains/research/ml/dataset/anonymizer/transformations/anonymization/JvmAnonymizationVisitor.kt

```diff
@@ -0,0 +1,47 @@
+package org.jetbrains.research.ml.dataset.anonymizer.transformations.anonymization
+
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiRecursiveElementVisitor
+import com.intellij.refactoring.rename.RenamePsiElementProcessor.forElement
+import com.intellij.util.IncorrectOperationException
+import java.util.logging.Logger
+
+open class JvmAnonymizationVisitor(file: PsiFile, val anonymizer: JvmElementAnonymizer) :
+    PsiRecursiveElementVisitor() {
+    private val project = file.project
+    protected val log = Logger.getLogger(javaClass.name)
+
+    override fun visitElement(element: PsiElement) {
+        anonymizer.registerElement(element)
+        super.visitElement(element)
+    }
+
+    fun applyAnonymization(psiElement: PsiElement) {
+        psiElement.accept(this)
+        performAllRenames()
+    }
+
+    // TODO: it does not work for implements something now: see Java tests: classes_and_methods/out_2.java
+    private fun performAllRenames() {
+        val renames = anonymizer.allRenames.map { renameElementDelayed(it.first, it.second) }
+        WriteCommandAction.runWriteCommandAction(project) {
+            renames.forEach { it() }
+        }
+    }
+
+    private fun renameElementDelayed(definition: PsiElement, newName: String): () -> Unit {
+        val processor = forElement(definition)
+        processor.prepareRenaming(definition, newName, mutableMapOf(definition to newName))
+        val references = processor.findReferences(definition, definition.useScope, false)
+        val usages = references.map { processor.createUsageInfo(definition, it, it.element) }.toTypedArray()
+        return {
+            try {
+                processor.renameElement(definition, newName, usages, null)
+            } catch (e: IncorrectOperationException) {
+                log.info("Cannot perform the renaming")
+            }
```

Comment:
I think, yes, because honestly, I don't know exactly why this error can occur. Looking at the realization of renaming, it's necessary for the element to be PsiNamedElement (I check it in JvmTypes.toRename), but they also said in comments that this error can occur if it's impossible to perform a renaming for any reason. So I'm not sure what is the right thing to do -- raise an exception or log it out.

Reply:
I got it, thanks! Please add information about the element in the log